Split stream rewrite

gustavnikolaj opened the pull request:
This is a proper breaking change to the API. The diff in the tests are probably the easiest way to show it.

This implementation add support for node 14 (issue #22) and work around the issues that arose by running the old code on node 14 (done in PR #23).

We have had some issues with the existing implementation of hijackresponse with regards to how it worked with backpressure and how nicely it played with other streams, but it mostly worked most of the time, and it definitely worked well enough not to be a problem. I've tried to solve it multiple times, but my mind breaks whenever I try to understand what is going on, and it's remarkably hard to draw down on paper.

This new approach attempt to solve all those problems, by drastically simplifying the API. In the callback function you pass to `hijackResponse` you now get two different streams rather than one.

Before you would get a stream (which we will call res), that was both a Readable and a Writable stream; The readable part is what the downstream middleware thought that it wrote to the response (the actual hijacked response) and the writeable part is the actual response that you can write the modified data to.

Now you instead get two streams: the first is a readable stream (actually, it's a Transform stream, but you should treat it as a Readable) where the response body of the hijacked request - the second one is the actual writable request.

The issue that arose for node 14 was that calling unpipe() on a stream will now also implicitly call pause() - and due to the way we monkeypatched the Readable-stream onto the `res` object, the two streams would now both react to some of those events (e.g. pause). That was caught by a test in our test suite.

I was able to think of two solutions:

1) Filter which events we relayed to the Readable-stream from res - having potential for all sort of other bugs and misbehavior - and entirely cutting us off from ever addressing issues like backpressure. 
2) Properly separate the two streams.

This PR is my take on the latter.

## API Changes

The current readme example looks like this:

```js
app.use(function(req, res, next) {
  hijackResponse(res, function(err, res) {
    if (err) {
      res.unhijack(); // Make the original res object work again
      return next(err);
    }

    res.setHeader("X-Hijacked", "yes!");
    res.removeHeader("Content-Length");

    res.pipe(transformStream).pipe(res);
  });
  next();
});
```

This proposal would change it to:

```js
app.use(function(req, res, next) {
  hijackResponse(res, function(err, hijackedResponseBody, res) {
    if (err) {
      // err is never not null, can be changed in a later update.
      // But for the sake of the example, this is how you unhijack:
      hijackedResponseBody.pipe(res);
      return
    }

    res.setHeader("X-Hijacked", "yes!");
    res.removeHeader("Content-Length");

    hijackedResponseBody.pipe(transformStream).pipe(res);
  });
  next();
});
```

First an foremost - it's a simpler mental model - the res object inside your hijackResponse callback is now just the actual res, as it was before it was hijacked (**ALMOST**).

It also gets rid of a few weird things about the API - such as the `res.pipe(res)` weirdness and the `res.unhijack()` and `res.destroyHijacked()` constructs.

`res.pipe(res)` now just becomes `hijackedResponseBody.pipe(res)` - and if you don't have anything else in your pipe, it would effectively be the same as `res.unhijack()`.

`res.destroyHijacked()` is supposed to discard the body of the hijacked response, which would be replaced by `hijackedResponseBody.destroy()`.

**Edit:** I had to add `hijackedResponseBody.destroyAndRestore()` for some uses of `res.destroyHijacked()` and `res.unhijack()` in express-processimage. A combination of those two methods was used to discard the response so far, and restore `res` enough that calling `next()` with an error would work. `next` from express/final-handler also handles errors coming up after the headers has been sent, so just calling `hijackedResponseBody.destroy()` and restoring the patched methods is enough.

## Example Conversions

- https://github.com/papandreou/express-extractheaders/pull/40
- https://github.com/papandreou/express-processimage/pull/120
- https://github.com/papandreou/express-compiless/pull/46

## Drawbacks to this approach

### 1. Performance Overhead

In the case where you previously would call `res.unhijack()`, it would undo all the modifications to `res` and make the handlers that would have had their responses hijacked write directly to the response stream. With this change, you would, in that situation, now suffer the overhead of the downstream handlers writing through a PassThrough stream, as you would have to unhijack it by using `hijackedResponseBody.pipe(res)`.

This problem can be lessened by using a [minipass](https://github.com/isaacs/minipass) stream instead. We can also make it an optional performance flag.

gustav-olsen-groupone committed:
Start testing on node 14

gustav-olsen-groupone committed:
Hand out the hijacked response body as a stream

gustav-olsen-groupone committed:
Add Object.entries for node 6

gustav-olsen-groupone committed:
Drop support for node version older than 8

In node 8 `lib/_http_outgoing.js` was changed in node core. The current
implementation works against node 8 and above, but I have not
been able to diagnose exactly what breaks in the old version.

It seems to be something with res.write being bound in the wrong way -
at least, what happens is that the overwritten version of res.write is
called from within the original res.end.

Since node 6 has been end of lifed for almost a couple of years, I will
not spend further time debugging it and just bump the minimal required
node version... and schedule a code-style overhaul, since we are now
officially not supporting any non ES6-capable node versions.

gustav-olsen-groupone committed:
Add destoryAndRestore method to reponseBody stream

In express-processimage res.unhijack() was used as a way to discard the
stream and allow normal error handling by passing errors to `next` to
work again.

That was an unsafe/unsupported operation in the old implementation, and
it would at the very least result in dangling half-read streams.

This implementation, with this method added, should address any issues
with that usage pattern.

gustavnikolaj commented:
I had to add hijackedResponseBody.destroyAndRestore() for some uses of res.destroyHijacked() and res.unhijack() in express-processimage. A combination of those two methods was used to discard the response so far, and restore res enough that calling next() with an error would work. next from express/final-handler also handles errors coming up after the headers has been sent, so just calling hijackedResponseBody.destroy() and restoring the patched methods is enough.

papandreou commented:
This looks good to me. Now that we're doing a massive breaking change, I think we should also consider:

1. Not handing `res` to the code that hijacks -- if it's the same object, the hijacking code should already have access to it?
2. Making the `hijackResponse` function return a promise instead of being callback-oriented. If we go for 1. there's only a single value to resolve with anyway.

gustavnikolaj commented:
I’d love to do that, but unfortunately `res` in the callback is not identical to res on the outside. 

Calling `hijackResponse` overwrites some methods on the `res` reference. The `res` passed to the callback is an object with the original versions of the overwritten methods and with res as it’s prototype.

We need the res that is passed on to the next handler to be rewired to write to our responseBodyStream - and I don’t know how to do that without changing our reference to res aswell. Using a function allows us to shadow the monkey-patched res intended for the other handlers, with our own restored version of res (proxyRes in this implementation)

There’s no async setup in the `hijackResponse` function, so we don’t need a callback or a promise - we could just return the two streams immediately - but we would need to call the “inner-res” something else - `actuallyTheRealRes` - as they aren’t exactly equal. 

We are calling the callback in the patched version of res.writeHead because at that point you know that the handlers below you are done modifying headers and statuscode. You can get the same info by waiting for the `readable` event on the hijacked responseBody stream or wait for the first chunk. 

I am all for suggestions towards improving the API - and we can try to add different interfaces as well to see what works best in practice. At the very least im going to remove the err arg from the callback :-) I just wanted to focus on this one change for this PR - it's plenty complex by itself. But it opens the door for many new changes as the implementation (in my humble opinion) is much simpler now.

papandreou commented:
> Calling `hijackResponse` overwrites some methods on the `res` reference. The `res` passed to the callback is an object with the original versions of the overwritten methods and with res as it’s prototype.

Ah, right, I had completely forgotten about all that 🙈

> We need the res that is passed on to the next handler to be rewired to write to our responseBodyStream - and I don’t know how to do that without changing our reference to res aswell. Using a function allows us to shadow the monkey-patched res intended for the other handlers, with our own restored version of res (proxyRes in this implementation)

Hmm, yeah, I think I looked into that but found that express closes over `res`, so it wasn't possible to get it to hand a different object out to the other handlers. Dammit.

gustavnikolaj commented:
As for alternative APIs my imagination does not go further than the following example - at least at the time.

A sync API, where you have to wait until looking at / setting headers:

```js
app.use((req, res, next) => {
  const {responseBody, actualRes } = hijackResponse(res);

  responseBody.once('readable', () => {
    actualRes.setHeader("X-Hijacked", "yes!");
    actualRes.removeHeader("Content-Length");
  });

  responseBody.pipe(transformStream).pipe(actualRes);

  next();
});
```

I actually think I prefer what we have now, but improved to not include the `err` argument.

```js
app.use((req, res, next) => {
  hijackResponse(res, (responseBody, res) => {
    res.setHeader("X-Hijacked", "yes!");
    res.removeHeader("Content-Length");

    responseBody.pipe(transformStream).pipe(res);
  });
  next();
});
```

Either way, I will move forward with this pull-request and do a little follow-up work before releasing a new major. I'll send PRs for the updates. :-) We can always go back to discussing the API - we have a better chance to improve it now. And now we have a version that is actually safe with recent node too! ;-)

gustav-olsen-groupone committed:
Proxy statusCode and statusMessage to res

This is not happening as we are using an empty object with res as it's
proxy type. It would be really cool to be able to use a Proxy here, but
I will have to spend a little more time getting that right.

A proxy will be much more stable than this sort of hackery.

gustavnikolaj commented:
CI failure is only due to coverage decrease.

gustavnikolaj commented:
Released in 5.0.0-rc.1 under the next dist-tag for now.

alexjeffburke commented:
@gustavnikolaj I think you were right to move forward with it :)

I did try to have a brief look, but there wasn't really anything I felt I could comment on, seemed like you'd put a lot of thought into the API. There were a couple of removals that caused me a raised eyebrow at first, but think I convinced myself they were fine given the rethink - like not setting the prototype on the handed in responseBody which I assume is because there is no need to pretend any more). Re the `res` now given to the callback, I know it's instrumented so the name makes me a bit nervous because it implies it is exactly the same as outside, but I assume the outside one is affected by the monkey patchingtoo  so likely it doesn't matter. If I had carte blanche I might call them `proxiedResponseBody` and `wrappedRes` or something, but I've no strong feelings so go with your gut.

gustavnikolaj commented:
The naming of the args are up to the user, but I agree that it is really important. If nothing else, we should be clear in what we call them internally.

Regarding `proxiedResponseBody` I used `hijackedResponseBody` in the updated README example. I think that is a bit more precise than proxied, but I'll happily hear arguments to convince me otherwise. I have used a shorter variant in the code that I wrote for express-processimage etc - `responseBody`. I think it is pretty clear what it is either way, as it is the whole point of using the hijackresponse module in the first place.

I do, however, still prefer res as the name in the callers callback. The `wrapped` res is really the one that you got from express - that is the one we have monkeypatched to write into our responseBodyStream instead. The `res` that you're passed in the callback is the one that will be writing to the actual response :-)

I call the `res` which is passed to the callback `proxyRes` in the code. It's not really a good name, but it needed to have a name that wasn't `res` (as that is what the original res is called). Maybe it should be called restoredRes?

---

> I did try to have a brief look, but there wasn't really anything I felt I could comment on, seemed like you'd put a lot of thought into the API. There were a couple of removals that caused me a raised eyebrow at first, but think I convinced myself they were fine given the rethink - like not setting the prototype on the handed in responseBody which I assume is because there is no need to pretend any more).

>> [..] like not setting the prototype on the handed in responseBody [...]

This is actually specifically what I was going for. I wanted to split it out to undo the inverse-duplex-stream-nature of our butchered `res`. `res` in the old API would have a ReadableStream interface, which presented the data that other people thought they were writing, and a WritableStream interface, which would be what you were going to be writing.

What we have now, is something we can much more easily reason about - we have the `hijackedResponseBody` stream, which is just that - the hijacked response body - and then we have `proxyRes` which is really just the monkey-patched res, but restored to it's original pre-monkey-patching-glory.

What specifically prompted this change was that node 14 has begun to emit pause events on both sides when you unpipe. This would cause hijackresponse to hang indefinitely. It was making me insane trying to reason about it - but my best attempt at an explanation is that the events emitted on `res` actually went to both the readable and the writeable sides, when it was only intended to go to either. This split of the two streams means that there shouldn't be any ambiguity anymore.

Also, the implementation became a lot less intertwined. We basically just take control over the writable-stream methods on res, and patch them forward to our hijackedResponseStream, and then undo that for the `proxyRes` that we pass to the hijacker.

--- 

That was a lot of words, but I hope it addressed the feedback that you had. I am very happy that you took time to look over it, and that you don't think I messed it up completely :-)

gustavnikolaj commented:
Also, I appreciate that this is a really tough review to do. It's a really obscure piece of code, and it takes me a couple of days to really get it each time I pick it up.

papandreou commented:
I still think it would be worth it to try to get rid of the callback one way or the other. I guess that really would need to happen before 5.0.0, as it'd be pretty douchy to break backwards compatibility twice :)
